drivers/usb/udc/udc_nrf.c

```diff
@@ -0,0 +1,817 @@
+	struct net_buf *buf;
+
+	buf = udc_ctrl_alloc(dev, USB_CONTROL_EP_OUT, length);
+	if (buf == NULL) {
+		return -ENOMEM;
+	}
+
+	net_buf_put(&cfg->fifo, buf);
+	udc_nrf_clear_control_out(dev);
+
+	return 0;
+}
+
+static int udc_event_xfer_setup(const struct device *dev)
+{
+	struct net_buf *buf;
+	int err;
+
+	buf = udc_ctrl_alloc(dev, USB_CONTROL_EP_OUT,
+			     sizeof(struct usb_setup_packet));
+	if (buf == NULL) {
+		LOG_ERR("Failed to allocate for setup");
+		return -ENOMEM;
+	}
+
+	udc_ep_buf_set_setup(buf);
+	nrfx_usbd_setup_get((nrfx_usbd_setup_t *)buf->data);
+	net_buf_add(buf, sizeof(nrfx_usbd_setup_t));
+	udc_nrf_setup_rcvd = true;
+
+	/* Update to next stage of control transfer */
+	udc_ctrl_update_stage(dev, buf);
+
+	if (udc_ctrl_stage_is_data_out(dev)) {
+		/*  Allocate and feed buffer for data OUT stage */
+		LOG_DBG("s:%p|feed for -out-", buf);
+		err = usbd_ctrl_feed_dout(dev, udc_data_stage_length(buf));
+		if (err == -ENOMEM) {
+			err = udc_submit_event(dev, UDC_EVT_EP_REQUEST, err, buf);
+		}
+	} else if (udc_ctrl_stage_is_data_in(dev)) {
+		err = udc_ctrl_submit_s_in_status(dev);
+	} else {
+		err = udc_ctrl_submit_s_status(dev);
+	}
+
+	return err;
+}
+
+static void udc_nrf_thread(const struct device *dev)
+{
+	nrfx_usbd_evt_t event;
+	uint8_t ep;
+
+	while (true) {
+		k_msgq_get(&drv_msgq, &event, K_FOREVER);
+		ep = event.data.eptransfer.ep;
+
+		switch (event.type) {
+		case NRFX_USBD_EVT_EPTRANSFER:
+			if (USB_EP_DIR_IS_IN(ep)) {
+				udc_event_xfer_in(dev, &event);
+			} else {
+				udc_event_xfer_out(dev, &event);
+			}
+			break;
+		case NRFX_USBD_EVT_SETUP:
+			udc_event_xfer_setup(dev);
+			break;
+		default:
+			break;
+		}
+	}
+}
+
+static void udc_sof_check_iso_out(const struct device *dev)
+{
+	const uint8_t iso_out_addr = 0x08;
+	nrfx_usbd_evt_t event = {
+		.type = NRFX_USBD_EVT_EPTRANSFER,
+		.data = {
+			.eptransfer = {
+				.ep = iso_out_addr,
+				.status = NRFX_USBD_EP_WAITING,
+			},
+		},
+	};
+	struct udc_ep_config *ep_cfg;
+
+	ep_cfg = udc_get_ep_cfg(dev, iso_out_addr);
+	if (ep_cfg == NULL) {
+		return;
+	}
+
+	if (ep_cfg->stat.enabled && !k_fifo_is_empty(&ep_cfg->fifo)) {
+		k_msgq_put(&drv_msgq, &event, K_NO_WAIT);
+	}
+}
+
+static void usbd_event_handler(nrfx_usbd_evt_t const *const event)
+{
+	switch (event->type) {
+	case NRFX_USBD_EVT_SUSPEND:
+		LOG_INF("SUSPEND state detected");
+		nrfx_usbd_suspend();
+		udc_set_suspended(udc_nrf_dev, true);
+		udc_submit_event(udc_nrf_dev, UDC_EVT_SUSPEND, 0, NULL);
+		break;
+	case NRFX_USBD_EVT_RESUME:
+		LOG_INF("RESUMING from suspend");
+		udc_set_suspended(udc_nrf_dev, false);
+		udc_submit_event(udc_nrf_dev, UDC_EVT_RESUME, 0, NULL);
+		break;
+	case NRFX_USBD_EVT_WUREQ:
+		LOG_INF("Remote wakeup initiated");
+		break;
+	case NRFX_USBD_EVT_RESET:
+		LOG_INF("Reset");
+		udc_submit_event(udc_nrf_dev, UDC_EVT_RESET, 0, NULL);
+		break;
+	case NRFX_USBD_EVT_SOF:
+		LOG_DBG("SoF");
```

Comment:
Enabling debug level tracing essentially pollutes the log with SoF (every 1ms) and the logging becomes unusable. Please remove this line.